### PR TITLE
Add verify module

### PR DESCRIPTION
**Pull request by lkatalin**
A series of small changes leading to the addition of the `verify` module to `attestation_types`.

The OpenSSL methods that are going to be checking signatures on various parts of the Quote are going to need a `Vec<u8>` or `&[u8]` representation of specific inner structures, so methods are added to get these back out from a `Quote`. Some additional helpful changes are made to the `SigData` struct (the last part of a Quote holding useful data for verifying the Quote).

The `verify` module has some functions (in `cert_chain,rs`, `sig.rs`, `key.rs`) we're not using currently but might be helpful generally. I could also delete these. (They are from last year's demo.) It was complaining about dead code, so there's probably some better way to structure it to make that go away. I'll clean all this up more later anyway but I don't mind making changes to get this merged.

In this PR:
- Adding methods on `ECDSAP256Sig` and `ECDSAPKey` to get a vector of the structs
- Adding a `PCKCertChain` type
- Adding the cert data's length as a field to the `SigData` struct
- Changing the cert data type that `SigData` expects because this is what's in the V3 Quote
- Fixing some SigData` methods (suggestions welcome)
- Adding a method on `SigData` to get the cert data as a `PCKCertChain` type
- Adding methods on `report::Body` to get a vector of the struct
- Adding the verify module

Resolves https://github.com/enarx/enarx/issues/917
Resolves https://github.com/enarx/enarx/issues/789

### CODE REVIEW

File: src/attestation_types/quote/sigdata.rs [hunk 1]
```diff
@@ -31,6 +32,26 @@ impl From<&[u8; 64]> for ECDSAP256Sig {
...
+impl ECDSAP256Sig {
+    /// Returns r component
+    pub fn r(&self) -> &[u8; 32] {
+        &self.r
+    }
+
+    /// Returns s component
+    pub fn s(&self) -> &[u8; 32] {
+        &self.s
+    }
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:44]: `r` and `s` are already `pub` ... why these methods?

File: src/attestation_types/quote/sigdata.rs [hunk 2]
```diff
@@ -57,6 +78,56 @@ impl From<&[u8; 64]> for ECDSAPubKey {
...
+    /// Returns x coordinate
+    pub fn x(&self) -> &[u8; 32] {
+        &self.x
+    }
+
+    /// Returns y coordinate
+    pub fn y(&self) -> &[u8; 32] {
+        &self.y
+    }
+
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:91]: `x` and `y` are already `pub` ... why these methods?

File: src/attestation_types/quote/sigdata.rs [hunk 3]
```diff
@@ -57,6 +78,56 @@ impl From<&[u8; 64]> for ECDSAPubKey {
...
+    /// Returns intermediate certificate
+    pub fn intermed_cert(&self) -> &X509 {
+        &self.intermed_cert
+    }
+
+    /// Returns root certificate
+    pub fn root_cert(&self) -> &X509 {
+        &self.root_cert
+    }
+}
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:129]: These fields are also pub.

File: src/attestation_types/quote/sigdata.rs [hunk 4]
```diff
@@ -309,7 +311,9 @@ impl SigData {
     }
 
     /// Retrieve the QE Cert Data
-    pub fn get_qe_cert_data(&self) -> &Vec<u8> {
-        &self.qe_cert_data
+    pub fn qe_cert_data_vec(&self) -> Vec<u8> {
+        let mut v = Vec::new();
+        v.extend(&self.qe_cert_data);
+        v
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:317]:
```rust
        self.qe_cert_data.clone()
```

File: src/attestation_types/quote/sigdata.rs [hunk 5]
```diff
@@ -316,4 +316,29 @@ impl SigData {
         v.extend(&self.qe_cert_data);
         v
     }
+
+    /// Returns QE Cert Data as a PCK certificate chain, if QE Cert Data is of the appropriate type
+    pub fn qe_cert_data_pckchain(&self) -> Result<PckCertChain, Box<dyn Error>> {
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:321]:
```
    pub fn qe_cert_data_pckchain(&self) -> Result<PckCertChain, QuoteError> {
```

File: src/attestation_types/quote/sigdata.rs [hunk 6]
```diff
@@ -316,4 +316,29 @@ impl SigData {
...
+                "cannot return cert data as PCK cert chain; cert data is not PCK cert chain type"
+                    .to_string(),
+            )));
+        }
+
+        // Necessary because the formatting embedded in the Quote is incorrect. Fix later.
+        let chain = String::from_utf8(self.qe_cert_data_vec())?
+            .replace("-----END CERTIFICATE-----", "-----END CERTIFICATE-----\n");
+
+        let pck_cert_chain = X509::stack_from_pem(&chain.as_bytes().to_vec()[..])?;
```
Comment (haraldh) [inline src/attestation_types/quote/sigdata.rs:333]:
```rust
        let chain = String::from_utf8(self.qe_cert_data_vec())
            .map_err(|e| QuoteError(e.to_string()))?
            .replace("-----END CERTIFICATE-----", "-----END CERTIFICATE-----\n");

        let pck_cert_chain = X509::stack_from_pem(&chain.as_bytes().to_vec()[..])
            .map_err(|e| QuoteError(e.to_string()))?;
```

File: src/attestation_types/verify/mod.rs [hunk 7]
```diff
@@ -0,0 +1,114 @@
...
+        .verify_hash(hashed_reportdata, unhashed_data)?;
+    println!("CLIENT: 	 Enclave report hash OK");
+    println!("\nCLIENT: 	 Attestation Complete");
+
+    Ok(())
+}
+
+#[test]
+fn verify_sample_v3quote() {
+    assert!(!verify().is_err());
```
Comment (haraldh) [inline src/attestation_types/verify/mod.rs:113]:
```rust
    assert!(verify().is_ok());
```

**Review by haraldh — CHANGES_REQUESTED** [review]

**Comment by haraldh** [comment]
For `tokio` the minimum required version of rust has to be `1.45.0`.

So for the CI to succeed this is also needed:
```
diff --git a/.github/workflows/test.yml b/.github/workflows/test.yml
index 0675505..70cc3cd 100644
--- a/.github/workflows/test.yml
+++ b/.github/workflows/test.yml
@@ -18,7 +18,7 @@ jobs:
       fail-fast: false
       matrix:
         toolchain:
-          - 1.41.0
+          - 1.45.0
           - stable
           - beta
           - nightly
```

**Comment by haraldh** [comment]
I also would recommend something like
https://github.com/haraldh/sgx/commit/fe697310c5051bbf3e8f2ccbaff2411831bac451

to decouple the test suite from making online requests every time the test runs.

In the end we probably want something like `sgxctl export --full <filename>` to cache the cert chain.

**Comment by lkatalin** [comment]
> I also would recommend something like
> [haraldh@fe69731](https://github.com/haraldh/sgx/commit/fe697310c5051bbf3e8f2ccbaff2411831bac451)
> 
> to decouple the test suite from making online requests every time the test runs.
> 
> In the end we probably want something like `sgxctl export --full <filename>` to cache the cert chain.

@haraldh I've added these changes and added you as co-author. Also opened this issue to cache the cert chain: https://github.com/enarx/sgx/issues/18

**Comment by lkatalin** [comment]
@haraldh Thanks so much for these suggestions, they were extremely helpful. :+1: :+1: :+1:

File: src/attestation_types/quote/sigdata.rs (hunk 1 above)
Comment (lkatalin) [reply src/attestation_types/quote/sigdata.rs]: Good point, I removed all of these.

File: src/attestation_types/verify/mod.rs [hunk 8]
```diff
@@ -0,0 +1,129 @@
...
+
+    Ok(())
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    use sample_v3quote::{SAMPLE_INTEL_CERT_CHAIN, SAMPLE_V3QUOTE};
+
```
Comment (haraldh) [inline src/attestation_types/verify/mod.rs:118]: still needs:
```diff
diff --git a/src/attestation_types/verify/mod.rs b/src/attestation_types/verify/mod.rs
index 6b72593..bda2b24 100644
--- a/src/attestation_types/verify/mod.rs
+++ b/src/attestation_types/verify/mod.rs
@@ -114,7 +114,10 @@ pub fn verify(quote_bytes: &[u8], trusted_public_pck_chain: &str) -> Result<(),
 mod test {
     use super::*;
 
-    use sample_v3quote::{SAMPLE_INTEL_CERT_CHAIN, SAMPLE_V3QUOTE};
+    use sample_v3quote::SAMPLE_V3QUOTE;
+
+    #[cfg(not(feature = "chain_get"))]
+    use sample_v3quote::SAMPLE_INTEL_CERT_CHAIN;
 
     #[test]
     fn verify_sample_v3quote() {
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/attestation_types/verify/mod.rs (hunk 8 above)
Comment (lkatalin) [reply src/attestation_types/verify/mod.rs]: Fixed!

**Review by haraldh — DISMISSED** [review]

**Comment by lkatalin** [comment]
Changing to draft just because I want to include a few more tests for negative cases.

**Comment by lkatalin** [comment]
@haraldh I only made changes in a338d6641b30f3a8c225e85ae4472f8c6b577eaf .

File: src/attestation_types/verify/cert_chain.rs [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+            Ok(success) => {
+                if success {
+                    Ok(())
+                } else {
+                    Err(VerifyError(format!(
+                        "invalid signature in certificate chain: {}",
+                        context.error()
+                    )))
+                }
+            }
```
Comment (haraldh) [inline src/attestation_types/verify/cert_chain.rs:110]:
```suggestion
            Ok(true) => Ok(()),
            Ok(false) => Err(VerifyError(format!(
                "invalid signature in certificate chain: {}",
                context.error()
            ))),
```

File: src/attestation_types/verify/key.rs [hunk 10]
```diff
@@ -0,0 +1,171 @@
...
+            Ok(success) => {
+                if success {
+                    Ok(())
+                } else {
+                    Err(VerifyError("signature verification failed".to_string()))
+                }
+            }
+            Err(_) => Err(VerifyError(
+                "signature validity could not be determined".to_string(),
+            )),
```
Comment (haraldh) [inline src/attestation_types/verify/key.rs:155]:
```suggestion
            Ok(true) => Ok(()),
            Ok(false) => Err(VerifyError("signature verification failed".to_string())),
            Err(e) => Err(VerifyError(format!(
                "signature validity could not be determined: {}",
                e
            ))),
```

**Review by haraldh — CHANGES_REQUESTED** [review]

File: src/attestation_types/verify/cert_chain.rs (hunk 9 above)
Comment (lkatalin) [reply src/attestation_types/verify/cert_chain.rs]: You're right, this is better.

**Review by haraldh — APPROVED** [review]

File: src/attestation_types/verify/key.rs [hunk 11]
```diff
@@ -0,0 +1,167 @@
...
+
+/// This is the error returned when the PCK hash is not valid.
+#[derive(Debug, Clone)]
+pub struct HashError;
+
+impl Error for HashError {
+    fn source(&self) -> Option<&(dyn Error + 'static)> {
+        // Generic error, underlying cause isn't tracked.
+        None
+    }
```
Comment (connorkuehl) [inline src/attestation_types/verify/key.rs:26]:
```suggestion
```

This is already `std::error::Error`'s default implementation if an implementor does not supply their own.

File: src/attestation_types/verify/key.rs [hunk 12]
```diff
@@ -0,0 +1,167 @@
...
+            openssl::ec::PointConversionForm::UNCOMPRESSED,
+            &mut *new_ctx,
+        )?;
+
+        Ok(tenant_pubkey_bytes)
+    }
+
+    /// DHKE deriving shared secret between self's private key and peer's public key.
+    pub fn derive_shared_secret(&self, peer_key: &PKey<Public>) -> Result<Vec<u8>, Box<dyn Error>> {
+        let ec_priv_key = self.privkey.as_ref().unwrap();
```
Comment (connorkuehl) [inline src/attestation_types/verify/key.rs:133]: Is this `unwrap()` safe or should it be followed by a `.ok_or(...)?;`

File: src/attestation_types/verify/key.rs [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+            curve,
+            pubkey: pub_pkey,
+            privkey: None,
+        })
+    }
+
+    /// This creates a new Key from existing PKey value.
+    pub fn new_from_pubkey(pkey: PKey<Public>) -> Self {
+        Key {
+            curve: EcGroup::from_curve_name(Nid::X9_62_PRIME256V1).unwrap(),
```
Comment (connorkuehl) [inline src/attestation_types/verify/key.rs:88]: Is this unwrap safe?

**Review by connorkuehl — CHANGES_REQUESTED** [review]

File: src/attestation_types/verify/cert_chain.rs [hunk 14]
```diff
@@ -0,0 +1,112 @@
...
+        self.chain.len() + 1
+    }
+
+    /// This checks that the CertChain's length is under the maximum allowed.
+    pub fn len_ok(&self) -> Result<(), Box<dyn Error>> {
+        if self.chain.len() > self.max_len {
+            panic!("Certificate chain length exceeds max allowable.");
+        }
+        Ok(())
+    }
```
[the remark this answers is not in the document]
Comment (lkatalin) [reply src/attestation_types/verify/cert_chain.rs:45]: Ah, you're right. I think this should error out because of the kind of check it's meant to be (something we decided on last year about a max allowable chain length), but it shouldn't panic.

File: src/attestation_types/verify/key.rs (hunk 11 above)
Comment (lkatalin) [reply src/attestation_types/verify/key.rs:26]: You're right. Actually I'm removing this whole error because I made a more general one with `VerifyError`. This is one of those things that was left in the code from the other demo.

File: src/attestation_types/verify/key.rs (hunk 13 above)
Comment (lkatalin) [reply src/attestation_types/verify/key.rs]: It wasn't. Thanks.

**Comment by lkatalin** [comment]
Thanks for the good suggestions, @connorkuehl ! I reworked some of these functions and their error handling.

**Review by connorkuehl — APPROVED** [review]

**Review by haraldh — APPROVED** [review]
